### PR TITLE
STONITH for remote nodes

### PR DESCRIPTION
**Pull request by vuntz**
Please note that this depends (and contains) https://github.com/crowbar/crowbar-ha/pull/55

### CODE REVIEW

File: chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb [hunk 1]
```diff
@@ -144,6 +149,26 @@ def self.cluster_nodes(node, role = "*")
     end
   end
 
+  # Performs a Chef search and returns an Array of Node objects for
+  # all remote nodes in the same cluster as the given node, or an empty array
+  # if the node isn't in a cluster.  Can optionally include remote nodes that
+  # are will be part of the cluster but are not setup yet.
```
Comment (jsuchome) [inline chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb]: Typo: *are will be*

File: chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb [hunk 2]
```diff
@@ -144,6 +149,26 @@ def self.cluster_nodes(node, role = "*")
...
+  def self.remote_nodes(node, include_not_setup = false)
+    if cluster_enabled?(node)
```
Comment (jsuchome) [inline chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb]: `return [] unless cluster_enabled?(node)`

File: chef/cookbooks/pacemaker/recipes/sbd.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    action :nothing
+    subscribes :run, "cookbook_file[sbd_remote.service]", :immediately
+  end
+end
+
+service "sbd" do
+  service_name "sbd_remote" if node[:pacemaker][:is_remote]
+  action :enable
+  # with systemd (so no SLES11), sbd needs to be enabled
+  not_if { node[:platform] == "suse" && node[:platform_version].to_f < 12.0 }
```
Comment (jsuchome) [inline chef/cookbooks/pacemaker/recipes/sbd.rb]: Not that it really matters, but this code is for Cloud6 only, right? So this condition should not be met...

File: chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb (hunk 2 above)
Comment (vuntz) [inline chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb]: Are you fine if I keep that for a followup PR? We should do it for all methods, not just this new one.

Comment (jsuchome) [inline chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb]: Yes

File: chef/cookbooks/pacemaker/recipes/sbd.rb (hunk 3 above)
Comment (vuntz) [inline chef/cookbooks/pacemaker/recipes/sbd.rb]: The pacemaker cookbook is intended to have a life outside of crowbar (for OpenStack upstream, for instance), so I try to keep things compatible with SLE11 there.

File: chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb (hunk 1 above)
Comment (vuntz) [inline chef/cookbooks/crowbar-pacemaker/libraries/helpers.rb]: fixed

**Comment by jsuchome** [comment]
OK, starting with **pacemaker: Split out non-stonith sbd code to its own recipe** commit, I do not see any real problem. So LGTM

File: chef/cookbooks/crowbar-pacemaker/recipes/remote.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+if node[:pacemaker][:stonith][:mode] == "sbd"
+  include_recipe "crowbar-pacemaker::sbd"
+
+  stonith_node_name = "remote-#{node[:hostname}"
+  if node[:pacemaker][:stonith][:sbd][:nodes][node[:fqdn]][:slot_name] != stonith_node_name
+    node[:pacemaker][:stonith][:sbd][:nodes][node[:fqdn]][:slot_name] = stonith_node_name
+    node.save
+  end
+end
```
Comment (vuntz) [inline chef/cookbooks/crowbar-pacemaker/recipes/remote.rb]: I think this message is outdated.

**Comment by vuntz** [comment]
Rebased after the merge of the watchdog PR (that was conflicting with this).

**Comment by jsuchome** [comment]
Still +1. I haven't tested it on new installation, but I do not see anything wrong (and we could have new PR for potential new fixes)

File: chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
+#
+# Cookbook Name:: crowbar-pacemaker
+# Recipe:: stonith
+#
+# Copyright 2014, SUSE
```
Comment (aplanas) [inline chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb]: name of the recipe, date and `SUSE LINUX GmbH`? (also in chef/cookbooks/pacemaker/recipes/sbd.rb)

Comment (vuntz) [inline chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb]: I'm not working for `SUSE LINUX GmbH` :-) All my files are just SUSE.

File: chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+sbd_devices = nil
+sbd_devices ||= (node[:pacemaker][:stonith][:sbd][:nodes][node[:fqdn]][:devices] rescue nil)
+sbd_devices ||= (node[:pacemaker][:stonith][:sbd][:nodes][node[:hostname]][:devices] rescue nil)
+
+sbd_devices.each do |sbd_device|
+  if File.symlink?(sbd_device)
```
Comment (aplanas) [inline chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb]: is this for cases where we refer the device by IDs?

File: chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb (hunk 5 above)
Comment (vuntz) [inline chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb]: Fixed the name of the recipe

**Comment by vuntz** [comment]
Force-pushed: name of recipes in headers fixed.

**Comment by jsuchome** [comment]
(still +1)

File: chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb (hunk 6 above)
Comment (vuntz) [inline chef/cookbooks/crowbar-pacemaker/recipes/sbd.rb]: The user can refer to the device by any path, can be uuid, id, etc. What we want is to have the absolute real path of the device.

(note that this is not new code, but moved code, btw)

**Comment by aplanas** [comment]
+1 (not tested, but review the code)

**Comment by vuntz** [comment]
Gating was green before the last force-push, and that last force-push was only changing comments => merging.

**Pull request merged**